Ethiopian calendar

grzesiek2010 opened the pull request:
Closes #1330 

#### What has been done to verify that this works as intended?
I've tested attached form.

#### Why is this the best possible solution? Were any other approaches considered?
I can't see any better solution. I think it's clear and refactoring will allow us to add another calendars easily.

#### Are there any risks to merging this code? If so, what are they?
Yes this pr contains code refactoring so we need @mmarciniak90 to test everything carefully

[ethiopian.txt](https://github.com/opendatakit/collect/files/1345883/ethiopian.txt)

@jknightco @lognaturel I think you are interested in this.

grzesiek2010 committed:
Added resources

grzesiek2010 committed:
Created custom date picker dialog layout

grzesiek2010 committed:
Implemented EthiopianWidget

grzesiek2010 committed:
Added licenses

grzesiek2010 committed:
Code refactoring

grzesiek2010 committed:
Fixed problem with DateTimeWidget

grzesiek2010 committed:
Code refactoring

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Removed unused code

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Fixed pmd

grzesiek2010 committed:
Fixed years range and refactored the code

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Fixed bug with default values in year and month-year appearances

grzesiek2010 committed:
Fixed daylight saving gap issue in ethiopian calendar

lognaturel commented:
I got a chance to show this to an Ethiopian partner. He verified the day names in Amharic, in Tigrinia and anglicized and is happy with all. The big thing he checked was leap years and that's also working well. Nice work!

The one thing he asked for was for the verification Gregorian date underneath the date picker to always be in English, even when the device is in Amharic or Tigrinia. I think forcing a US locale for that date should do the trick I think.

grzesiek2010 commented:
@lognaturel 
Nice to hear that!

> The one thing he asked for was for the verification Gregorian date underneath the date picker to always be in English, even when the device is in Amharic or Tigrinia. I think forcing a US locale for that date should do the trick I think.

It's not always in English format since I used this method
https://github.com/grzesiek2010/collect/blob/ebebfb9848d63f0f3086a1096361e32d340293c8/collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java#L26

it should be easy to change but what about that date I display after Ethiopian date in brackets? Should it be always in English too? (it's in the widget not in dialog)

grzesiek2010 committed:
Code improvemnets

lognaturel commented:
Thanks for thinking of that. He says both should be in English.

Another thing -- I think the strings will be very confusing for translators. How about calling them `ethiopian-month-meskerem`, etc instead of just `meskerem`?

grzesiek2010 committed:
Use only English date format in ethiopian widget

grzesiek2010 committed:
Fixed strings naming

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
@lognaturel 
I fixed both issues. I think it's ready for review/testing.

What should be tested:
EthiopianDateWidget with possible appearances: "month-year", "year".
DateTimeWidget with EthiopianDateWidget.

DaylightSaving gaps - it was a problem in EthiopianDateWidget but I fixed it. Example data: 01-01-1960 in Nairobi.

I refactored the code so DateWidget should be tested too in order to avoid any regressions (possible appearances: "no-calendar", "month-year", "year".
I think the attached above form contains all possible widgets.

grzesiek2010 committed:
Fixed Ethiopian date label with nonth-year appearance

mmarciniak90 commented:
I noticed unexpected behavior in date and time question.
**Use case:**
1. User adds date and time
2. User clicks on 'Go To Prompt' icon
3. User sees question and added answer
4. Time is not visible next to Ethiopian date. Time is in brackets, next to US date.
![screenshot_2017-10-13-16-14-01](https://user-images.githubusercontent.com/3907554/31550795-cffbb4fe-b032-11e7-94fd-0620c974cb57.png)

**Expected result**
Time is visible next to both dates

grzesiek2010 committed:
Code refactoring: implemented DatePickerDetails class to deal with types and modes easier

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Fixed problem with time in ethiopian date

grzesiek2010 committed:
Added Copyright headers

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2953,6 +2956,11 @@ public void onNumberPickerValueSelected(int widgetId, int value) {
 
+    @Override
+    public void onDateChanged(LocalDateTime date) {
+        ((ODKView) currentView).setBinaryData(date);
```
Dangerous access of `currentView`. Can you wrap this in an `instanceof` block?

grzesiek2010 replied:
Fixed using https://github.com/grzesiek2010/collect/commit/e979e292bf285f318b1c5c7150cbc96e0f01775a

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/CustomDatePickerDialog.java:
```diff
@@ -0,0 +1,195 @@
+    private static final String DATE = "date";
+    private static final String DATE_PICKER_DETAILS = "datePickerDetails";
+
+    protected NumberPicker dayPicker;
```
Would prefer to avoid `protected` access when possible.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/CustomDatePickerDialog.java:
```diff
@@ -0,0 +1,195 @@
+        void onDateChanged(LocalDateTime date);
+    }
+
+    private CustomDatePickerDialogListener listener;
```
Should be with the rest of the attribute definitions above.

grzesiek2010 replied:
Fixed.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/CustomDatePickerDialog.java:
```diff
@@ -0,0 +1,195 @@
+    public void onAttach(Context context) {
+        super.onAttach(context);
+        try {
+            listener = (CustomDatePickerDialogListener) context;
```
`instanceof` should be faster than a try/catch block, and codes to the expected case instead of the unexpected one.

grzesiek2010 replied:
Fixed.

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/CustomDatePickerDialog.java:
```diff
@@ -0,0 +1,195 @@
+
+    protected static Bundle getArgs(FormIndex formIndex, LocalDateTime date, DatePickerDetails datePickerDetails) {
+        Bundle args = new Bundle();
+        args.putSerializable(FORM_INDEX, formIndex);
```
We should be careful about serializing the JavaRosa classes just because some of them feature some very deep nesting and we could end up serializing a lot more data than we need to. I think its *okay* here, as FormIndex looks simple enough, but its something we should watch out for.

Personally, I would add an `onPickerClick` method to the listener, and keep the `FormIndex` out of this class entirely, allowing the Widget to handle its own call to `setIndexWaitingForData`.

grzesiek2010 replied:
I think it's ok. FormIndex is used in the same way in other places as well as I remember and there is also a test: https://github.com/opendatakit/javarosa/commit/87fc3dcf69b55fa9548284e4a88e612527ae71a0#diff-00bf7c2c715f66829ca3844e22334112

lognaturel replied:
@grzesiek2010 is there a specific reason not to follow what @jknightco is recommending with `onPickerClick`? Perhaps you two could quickly discuss in #collect-code so we're making a conscious decision here. 😊 In general I agree that keeping `FormIndex` out would be a good thing.

grzesiek2010 committed:
Fixed bugs

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/widgets/AbstractDateWidget.java:
```diff
@@ -0,0 +1,158 @@
+
+    @Override
+    public void setBinaryData(Object answer) {
+        date = (LocalDateTime) answer;
```
Unverified cast.

grzesiek2010 replied:
Fixed

heyjamesknight commented on collect_app/src/main/java/org/odk/collect/android/widgets/AbstractDateWidget.java:
```diff
@@ -0,0 +1,158 @@
+    protected Button dateButton;
+    protected TextView dateTextView;
+
+    protected boolean nullAnswer;
```
Can you add a comment describing what this is?

heyjamesknight replied:
Or maybe just rename it `isAnswerNull`?

grzesiek2010 replied:
Fixed.

heyjamesknight reviewed (changes requested):
Minor formatting, safety suggestions. Definitely would prefer to do without public/protected non-final fields in all classes moving forward.

grzesiek2010 commented:
@mmarciniak90 that bug with time has been fixed.

grzesiek2010 commented:
Thanks @jknightco I fixed bugs you have found apart from this one: https://github.com/opendatakit/collect/pull/1505#discussion_r144790243 since I don't find it dangerous I think using protected fields is ok. If you really want to stop using protected accesses we should discusses it (not two of us but the others too).

heyjamesknight commented:
I'm fine with leaving them protected if you can add a comment as to **why** they're protected and what an overriding class is expected to do with them. When/why should they be set, what are the accepted values, what should the values be interpreted as, etc.

Bloch does a better job explaining it than I can in Effective Java with Item 17: Design and document for inheritance or else prohibit it.

mmarciniak90 commented:
Tested with success. Verified on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0
![screenshot_20171016-154221](https://user-images.githubusercontent.com/3907554/31616303-a71f2ce4-b28c-11e7-8ad2-b79e287b1cb3.png)
Time is visible on list with all questions:
![screenshot_20171016-154210](https://user-images.githubusercontent.com/3907554/31616324-b3be9f0c-b28c-11e7-98f4-47853bf0b5b0.png)
Verified Tigrinya (ti) and Amharic (am) languages:
![tigrinya](https://user-images.githubusercontent.com/3907554/31616363-d18a7b3c-b28c-11e7-9baf-7f36abef3771.png)
![tigrinya2](https://user-images.githubusercontent.com/3907554/31616369-d4a53316-b28c-11e7-9bbb-08913ce145d8.png)
![amharic](https://user-images.githubusercontent.com/3907554/31616373-d68dacd0-b28c-11e7-9829-41328e3f1797.png)
![amharic2](https://user-images.githubusercontent.com/3907554/31616375-d8c2faa0-b28c-11e7-87d2-4cd1fd1cec7f.png)

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

grzesiek2010 committed:
Improved test to check hour format

grzesiek2010 committed:
Fixed skipDaylightSavingGapIfExistsTest

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
@jknightco I've reviewed that class again and I made some changes, not because I'm not in favor of protected accesses :smile: I just noticed I can move some code to the abstract class. So now it looks like you expected I think.

heyjamesknight commented:
@grzesiek2010 this looks great, thanks for letting me know.

lognaturel commented on collect_app/src/androidTest/java/org/odk/collect/android/utilities/DateTimeUtilsTest.java:
```diff
@@ -0,0 +1,248 @@
+    @Test
+    public void getDateTimeLabelTest() {
+        long dateInMilliseconds = 687967200000L; // 20 Oct 1991 14:00
+        Locale.setDefault(new Locale("af"));
```
Are all of these different locale tests actually testing different cases? I think one Gregorian English, one Gregorian non-English, one Ethiopic Amharic and one Ethiopic English would be enough. This would show that dates are localized and that in the case of an Ethiopic date the Gregorian date in parens is always in English.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/ApplicationConstants.java:
```diff
@@ -32,7 +32,7 @@
             "hi", "hu", "in", "it", "ja", "ka", "km", "lo_LA", "lt", "mg", "mr",
             "my", "nb", "ne_NP", "nl", "no", "pl", "ps", "pt", "ro", "ru", "so",
-            "sq", "sw", "sw_KE", "ta", "th_TH", "tl", "tr", "uk", "ur",
+            "sq", "sw", "sw_KE", "ta", "th_TH", "ti", "tl", "tr", "uk", "ur",
```
Thanks for remembering to do this. I also added it in Transifex at https://www.transifex.com/opendatakit/collect/language/ti/

lognaturel commented on collect_app/src/main/res/values-am/strings.xml:
```diff
@@ -11,4 +11,17 @@
   <!--%1$s represents degrees, %2$s represents minutes and %3$s represents seconds as in E 18°48'8"-->
   <!--%1$s represents degrees, %2$s represents minutes and %3$s represents seconds as in S 18°48'8"-->
   <!--%1$s represents degrees, %2$s represents minutes and %3$s represents seconds as in N 18°48'8"-->
+  <string name="ethiopian_month_meskerem">መስከረም</string>
```
We'll need to update these in Transifex once the English file updates there. These files are all replaced by the latest from Transifex on release.

lognaturel commented:
Thanks @grzesiek2010 @mmarciniak90 @jknightco! Our Ethiopian users will be very happy!

I made a couple of small notes inline.
